In [1]:
import pandas as pd

In [2]:
from urllib.request import urlretrieve

italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url, 'italy-covid-daywise.csv')

('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x275ceb61b80>)

In [3]:
covid_df = pd.read_csv('italy-covid-daywise.csv')

In [4]:
type(covid_df)

pandas.core.frame.DataFrame

In [5]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [6]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [7]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

In [8]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

Here's a summary of the functions & methods we've looked at so far:

* `pd.read_csv` - Read data from a CSV file into a Pandas `DataFrame` object
* `.info()` - View basic infomation about rows, columns & data types
* `.describe()` - View statistical information about numeric columns
* `.columns` - Get the list of column names
* `.shape` - Get the number of rows & columns as a tuple

# Retriving data from a data frame

In [9]:
# Pandas format is simliar to this
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

Format is of dictionary

Representing data in the above format has a few benefits:

* All values in a column typically have the same type of value, so it's more efficient to store them in a single array.
* Retrieving the values for a particular row simply requires extracting the elements at a given index from each column array.
* The representation is more compact (column names are recorded only once) compared to other formats that use a dictionary for each row of data (see the example below).

In [10]:
covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

Each column is represented using a data structure called `Series`, which is essentially a numpy array with some extra methods and properties.

In [11]:
type(covid_df['new_cases'])

pandas.core.series.Series

In [12]:
covid_df['new_cases'][246]

975.0

Pandas also provides the `.at` method to retrieve the element at a specific row & column directly.

In [13]:
covid_df.at[246, 'new_cases']

975.0

Instead of using the indexing notation `[]`, Pandas also allows accessing columns as properties of the dataframe using the `.` notation. However, this method only works for columns whose names do not contain spaces or special characters.

In [14]:
covid_df.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

# Subset from data

Further you can also pass a list of column within the indexing notation [] to access a subset of data frame with just the given columns

In [15]:
cases_df = covid_df[['date', 'new_cases']]

In [16]:
cases_df

date  new_cases
0    2019-12-31        0.0
1    2020-01-01        0.0
2    2020-01-02        0.0
3    2020-01-03        0.0
4    2020-01-04        0.0
..          ...        ...
243  2020-08-30     1444.0
244  2020-08-31     1365.0
245  2020-09-01      996.0
246  2020-09-02      975.0
247  2020-09-03     1326.0

[248 rows x 2 columns]

In [17]:
covid_df_copy = covid_df.copy()

To access a specific row of data, Pandas provides the `.loc` method.

In [18]:
covid_df.loc[243]

date          2020-08-30
new_cases         1444.0
new_deaths           1.0
new_tests        53541.0
Name: 243, dtype: object

Each retrived row is also a `Series` object

In [19]:
type(covid_df.loc[243])

pandas.core.series.Series

# First_valid_index() 

In [20]:
covid_df.new_tests.first_valid_index()

111

### Range of location

In [21]:
covid_df.loc[108 : 113]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0

We can use the `.sample` method to retrieve a random sample of rows from the data frame.

In [22]:
covid_df.sample(10)

date  new_cases  new_deaths  new_tests
140  2020-05-19      451.0        99.0    40226.0
179  2020-06-27      255.0        30.0    29721.0
36   2020-02-05        0.0         0.0        NaN
58   2020-02-27       78.0         1.0        NaN
196  2020-07-14      169.0        13.0    24222.0
145  2020-05-24      669.0       119.0    34206.0
116  2020-04-25     3021.0       420.0    38676.0
84   2020-03-24     4789.0       601.0        NaN
131  2020-05-10     1083.0       194.0    31384.0
150  2020-05-29      593.0        70.0    38233.0

Here's a summary of the functions & methods we looked at in this section:

- `covid_df['new_cases']` - Retrieving columns as a `Series` using the column name
- `new_cases[243]` - Retrieving values from a `Series` using an index
- `covid_df.at[243, 'new_cases']` - Retrieving a single value from a data frame
- `covid_df.copy()` - Creating a deep copy of a data frame
- `covid_df.loc[243]` - Retrieving a row or range of rows of data from the data frame
- `head`, `tail`, and `sample` - Retrieving multiple rows of data from the data frame
- `covid_df.new_tests.first_valid_index` - Finding the first non-empty index in a series


## Querying and sorting rows

Let's say we want only want to look at the days which had more than 1000 reported cases. We can use a boolean expression to check which rows satisfy this criterion.

In [23]:
high_new_cases = covid_df.new_cases > 1000

high_new_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

In [24]:
covid_df[high_new_cases]

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

We can write this succinctly on a single line by passing the boolean expression as an index to the data frame.

In [25]:
high_cases_df = covid_df[covid_df.new_cases > 1000]

high_cases_df

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

The data frame contains 72 rows, but only the first & last five rows are displayed by default with Jupyter for brevity. We can change some display options to view all the rows.

In [26]:
from IPython.display import display
with pd.option_context('display.max_rows' , 100):
    display(covid_df[covid_df.new_cases > 1000])

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
74   2020-03-14     2547.0       252.0        NaN
75   2020-03-15     3497.0       173.0        NaN
76   2020-03-16     2823.0       370.0        NaN
77   2020-03-17     4000.0       347.0        NaN
78   2020-03-18     3526.0       347.0        NaN
79   2020-03-19     4207.0       473.0        NaN
80   2020-03-20     5322.0       429.0        NaN
81   2020-03-21     5986.0       625.0        NaN
82   2020-03-22     6557.0       795.0        NaN
83   2020-03-23     5560.0       649.0        NaN
84   2020-03-24     4789.0       601.0        NaN
85   2020-03-25     5249.0       743.0        NaN
86   2020-03-26     5210.0       685.0        NaN
87   2020-03-27     6153.0       660.0        NaN
88   2020-03-28     5959.0       971.0        NaN
89   2020-03-29     5974.0       887.0        NaN
90   2020-03-30     5217.0       758.0        NaN
91   2020-03-31     4050.0       810.0        NaN
92   2020-04-01     4053.0       839.0        NaN
93   2020-04-02     4782.0       727.0        NaN
94   2020-04-03     4668.0       760.0        NaN
95   2020-04-04     4585.0       764.0        NaN
96   2020-04-05     4805.0       681.0        NaN
97   2020-04-06     4316.0       527.0        NaN
98   2020-04-07     3599.0       636.0        NaN
99   2020-04-08     3039.0       604.0        NaN
100  2020-04-09     3836.0       540.0        NaN
101  2020-04-10     4204.0       612.0        NaN
102  2020-04-11     3951.0       570.0        NaN
103  2020-04-12     4694.0       619.0        NaN
104  2020-04-13     4092.0       431.0        NaN
105  2020-04-14     3153.0       564.0        NaN
106  2020-04-15     2972.0       604.0        NaN
107  2020-04-16     2667.0       578.0        NaN
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
115  2020-04-24     2646.0       464.0    95273.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
119  2020-04-28     1739.0       333.0    37554.0
120  2020-04-29     2091.0       382.0    38589.0
121  2020-04-30     2086.0       323.0    41441.0
122  2020-05-01     1872.0       285.0    43732.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
126  2020-05-05     1221.0       195.0    32211.0
127  2020-05-06     1075.0       236.0    37771.0
128  2020-05-07     1444.0       369.0    13665.0
129  2020-05-08     1401.0       274.0    45428.0
130  2020-05-09     1327.0       243.0    36091.0
131  2020-05-10     1083.0       194.0    31384.0
134  2020-05-13     1402.0       172.0    37049.0
236  2020-08-23     1071.0         3.0    47463.0
237  2020-08-24     1209.0         7.0    33358.0
240  2020-08-27     1366.0        13.0    57640.0
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

## Analyzing data from data framesa
**Q: What are the total number of reported cases and deaths related to Covid-19 in Italy?**


In [28]:
total_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()

In [29]:
print('The number of reportted cases is {} and number of reported death is {}'.format(int(total_cases), int(total_deaths)))

The number of reportted cases is 271515 and number of reported death is 35497


**What is the overall death rate (ratoi of reported death to reported cases ?**

In [33]:
death_rate = total_deaths / total_cases

print("The overall death rate is {:.2f}%".format(death_rate*100))

The overall death rate is 13.07%


**What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were reported.**

In [34]:
initial_tests = 935310

total_tests = initial_tests + covid_df.new_tests.sum()

In [35]:
total_tests

5214766.0

**What fraction of tests returned a positive result?**

In [36]:
positive_rate = total_cases / total_tests

In [37]:
print('The positive rate is {:.2f}%'.format(positive_rate * 100))

The positive rate is 5.21%


We can also formulate more complex queries that involve multiple columns. As an example, let's try to determine the days when the ratio of cases reported to tests conducted is higher than the overall `positive_rate`.

In [38]:
positive_rate

0.05206657403227681

In [39]:
high_ratio_df = covid_df[covid_df.new_cases / covid_df.new_tests > positive_rate]

In [40]:
high_ratio_df

date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
120  2020-04-29     2091.0       382.0    38589.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
128  2020-05-07     1444.0       369.0    13665.0

In [41]:
covid_df.new_cases / covid_df.new_tests

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
243    0.026970
244    0.032055
245    0.018311
246         NaN
247         NaN
Length: 248, dtype: float64

we can use this series to add new column to the data frame

In [42]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests

In [43]:
covid_df.tail()

date  new_cases  new_deaths  new_tests  positive_rate
243  2020-08-30     1444.0         1.0    53541.0       0.026970
244  2020-08-31     1365.0         4.0    42583.0       0.032055
245  2020-09-01      996.0         6.0    54395.0       0.018311
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

for now lets remove `positive rate` column usind `drop` method

In [47]:
covid_df.drop(columns= ['positive_rate'], inplace= True)

In [49]:
covid_df.tail()

date  new_cases  new_deaths  new_tests
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

### Sorting rows using column values

The rows can also be sorted by a specific column using `.sort_values`. Let's sort to identify the days with the highest number of cases, then chain it with the `head` method to list just the first ten results.

In [51]:
covid_df.sort_values('new_cases', ascending= False).head()

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN

It looks like the last two weeks of March had the highest number of daily cases. Let's compare this to the days where the highest number of deaths were recorded.

In [56]:
covid_df.sort_values('new_deaths', ascending=False).head()

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN

It appears that daily deaths hit a peak just about a week after the peak in daily new cases.

Let's also look at the days with the least number of cases. We might expect to see the first few days of the year on this list.

# Removing error value

In [53]:
covid_df.sort_values('new_cases').head(10)

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
0    2019-12-31        0.0         0.0        NaN
29   2020-01-29        0.0         0.0        NaN
30   2020-01-30        0.0         0.0        NaN
32   2020-02-01        0.0         0.0        NaN
33   2020-02-02        0.0         0.0        NaN
34   2020-02-03        0.0         0.0        NaN
36   2020-02-05        0.0         0.0        NaN
37   2020-02-06        0.0         0.0        NaN
38   2020-02-07        0.0         0.0        NaN

It seems like the count of new cases on Jun 20, 2020, was `-148`, a negative number! Not something we might have expected, but that's the nature of real-world data. It could be a data entry error, or the government may have issued a correction to account for miscounting in the past. Can you dig through news articles online and figure out why the number was negative?

Let's look at some days before and after Jun 20, 2020.

In [55]:
covid_df.loc[169 : 175]

date  new_cases  new_deaths  new_tests
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20     -148.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

For now, let's assume this was indeed a data entry error. We can use one of the following approaches for dealing with the missing or faulty value:
1. Replace it with `0`.
2. Replace it with the average of the entire column
3. Replace it with the average of the values on the previous & next date
4. Discard the row entirely

Which approach you pick requires some context about the data and the problem. In this case, since we are dealing with data ordered by date, we can go ahead with the third approach.

You can use the `.at` method to modify a specific value within the dataframe.

In [57]:
covid_df.at[172, 'new_cases'] = (covid_df.at[171, 'new_cases'] + covid_df.at[173, 'new_cases']) / 2

In [59]:
covid_df.loc[171 : 173]

date  new_cases  new_deaths  new_tests
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20      297.5        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0

Here's a summary of the functions & methods we looked at in this section:

- `covid_df.new_cases.sum()` - Computing the sum of values in a column or series
- `covid_df[covid_df.new_cases > 1000]` - Querying a subset of rows satisfying the chosen criteria using boolean expressions
- `df['pos_rate'] = df.new_cases/df.new_tests` - Adding new columns by combining data from existing columns
- `covid_df.drop('positive_rate')` - Removing one or more columns from the data frame
- `sort_values` - Sorting the rows of a data frame using column values
- `covid_df.at[172, 'new_cases'] = ...` - Replacing a value within the data frame

# Working with dates

In [62]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

The data type of date is currently `object`, so Pandas does not know that this column is a date. We can convert it into a `datetime` column using the `pd.to_datetime` method.

In [64]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [65]:
covid_df.date

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

In [67]:
covid_df['year']     = pd.DatetimeIndex(covid_df.date).year
covid_df['month']    = pd.DatetimeIndex(covid_df.date).month
covid_df['day']      = pd.DatetimeIndex(covid_df.date).day
covid_df['weekdays'] = pd.DatetimeIndex(covid_df.date).weekday

In [69]:
covid_df.head()

date  new_cases  new_deaths  new_tests  year  month  day  weekdays
0 2019-12-31        0.0         0.0        NaN  2019     12   31         1
1 2020-01-01        0.0         0.0        NaN  2020      1    1         2
2 2020-01-02        0.0         0.0        NaN  2020      1    2         3
3 2020-01-03        0.0         0.0        NaN  2020      1    3         4
4 2020-01-04        0.0         0.0        NaN  2020      1    4         5

Let's check the overall metrics for May. We can query the rows for May, choose a subset of columns, and use the `sum` method to aggregate each selected column's values.

In [72]:
# Query for the rows in may
covid_df_may = covid_df[covid_df.month == 5]

# Extract the subset of columns to be aggregated
covid_df_may_metrics = covid_df_may[['new_cases', 'new_deaths', 'new_tests']]

# Get the column wise sum
covid_may_totals = covid_df_may_metrics.sum()

In [73]:
covid_may_totals

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

We can also combine the above operations into a single statement.

In [74]:
covid_df[covid_df.month == 5][['new_cases', 'new_deaths', 'new_tests']].sum()

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

## Grouping and aggregation

As a next step, we might want to summarize the day-wise data and create a new dataframe with month-wise data. We can use the `groupby` function to create a group for each month, select the columns we wish to aggregate, and aggregate them using the `sum` method. 

In [75]:
covid_month_df = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].sum()

In [76]:
covid_month_df

new_cases  new_deaths  new_tests
month                                  
1            3.0         0.0        0.0
2          885.0        21.0        0.0
3       100851.0     11570.0        0.0
4       101852.0     16091.0   419591.0
5        29073.0      5658.0  1078720.0
6         8217.5      1404.0   830354.0
7         6722.0       388.0   797692.0
8        21060.0       345.0  1098704.0
9         3297.0        20.0    54395.0
12           0.0         0.0        0.0

The result is a new data frame that uses unique values from the column passed to `groupby` as the index. Grouping and aggregation is a powerful method for progressively summarizing data into smaller data frames.

Instead of aggregating by sum, you can also aggregate by other measures like mean. Let's compute the average number of daily new cases, deaths, and tests for each month.

In [86]:
covid_df_mean = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].mean()

In [87]:
covid_df_mean

new_cases  new_deaths     new_tests
month                                       
1         0.096774    0.000000           NaN
2        30.517241    0.724138           NaN
3      3253.258065  373.225806           NaN
4      3395.066667  536.366667  38144.636364
5       937.838710  182.516129  34797.419355
6       273.916667   46.800000  27678.466667
7       216.838710   12.516129  25732.000000
8       679.354839   11.129032  35442.064516
9      1099.000000    6.666667  54395.000000
12        0.000000    0.000000           NaN

Apart from grouping, another form of aggregation is the running or cumulative sum of cases, tests, or death up to each row's date. We can use the `cumsum` method to compute the cumulative sum of a column as a new series. Let's add three new columns: `total_cases`, `total_deaths`, and `total_tests`.

In [91]:
covid_df['total_cases'] = covid_df.new_cases.cumsum()
covid_df['total_deaths']= covid_df.new_deaths.cumsum()
covid_df['total_tests'] = covid_df.new_deaths.cumsum()

In [92]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekdays  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31         1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1         2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2         3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3         4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4         5   
..         ...        ...         ...        ...   ...    ...  ...       ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30         6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31         0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1         1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2         2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3         3   

     total_cases  total_deaths  total_tests  
0            0.0           0.0          0.0  
1            0.0           0.0          0.0  
2            0.0           0.0          0.0  
3            0.0           0.0          0.0  
4            0.0           0.0          0.0  
..           ...           ...          ...  
243     267298.5       35473.0      35473.0  
244     268663.5       35477.0      35477.0  
245     269659.5       35483.0      35483.0  
246     270634.5       35491.0      35491.0  
247     271960.5       35497.0      35497.0  

[248 rows x 11 columns]

## Merging data from multiple sources

To determine other metrics like test per million, cases per million, etc., we require some more information about the country, viz. its population. Let's download another file `locations.csv` that contains health-related information for many countries, including Italy.

In [93]:
urlretrieve('https://gist.githubusercontent.com/aakashns/8684589ef4f266116cdce023377fc9c8/raw/99ce3826b2a9d1e6d0bde7e9e559fc8b6e9ac88b/locations.csv', 
            'locations.csv')
locations_df = pd.read_csv('locations.csv')

locations_df

location continent    population  life_expectancy  \
0      Afghanistan      Asia  3.892834e+07            64.83   
1          Albania    Europe  2.877800e+06            78.57   
2          Algeria    Africa  4.385104e+07            76.88   
3          Andorra    Europe  7.726500e+04            83.73   
4           Angola    Africa  3.286627e+07            61.15   
..             ...       ...           ...              ...   
207          Yemen      Asia  2.982597e+07            66.12   
208         Zambia    Africa  1.838396e+07            63.89   
209       Zimbabwe    Africa  1.486293e+07            61.49   
210          World       NaN  7.794799e+09            72.58   
211  International       NaN           NaN              NaN   

     hospital_beds_per_thousand  gdp_per_capita  
0                         0.500        1803.987  
1                         2.890       11803.431  
2                         1.900       13913.839  
3                           NaN             NaN  
4                           NaN        5819.495  
..                          ...             ...  
207                       0.700        1479.147  
208                       2.000        3689.251  
209                       1.700        1899.775  
210                       2.705       15469.207  
211                         NaN             NaN  

[212 rows x 6 columns]

In [94]:
locations_df[locations_df.location == "Italy"]

location continent  population  life_expectancy  \
97    Italy    Europe  60461828.0            83.51   

    hospital_beds_per_thousand  gdp_per_capita  
97                        3.18       35220.084

In [97]:
covid_df['location'] = "Italy"

covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekdays  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31         1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1         2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2         3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3         4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4         5   
..         ...        ...         ...        ...   ...    ...  ...       ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30         6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31         0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1         1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2         2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3         3   

     total_cases  total_deaths  total_tests location  
0            0.0           0.0          0.0    Italy  
1            0.0           0.0          0.0    Italy  
2            0.0           0.0          0.0    Italy  
3            0.0           0.0          0.0    Italy  
4            0.0           0.0          0.0    Italy  
..           ...           ...          ...      ...  
243     267298.5       35473.0      35473.0    Italy  
244     268663.5       35477.0      35477.0    Italy  
245     269659.5       35483.0      35483.0    Italy  
246     270634.5       35491.0      35491.0    Italy  
247     271960.5       35497.0      35497.0    Italy  

[248 rows x 12 columns]

We can now add the columns from `locations_df` into `covid_df` using the `.merge` method.

In [98]:
merge_df = covid_df.merge(locations_df, on = 'location')

In [100]:
merge_df.head()

date  new_cases  new_deaths  new_tests  year  month  day  weekdays  \
0 2019-12-31        0.0         0.0        NaN  2019     12   31         1   
1 2020-01-01        0.0         0.0        NaN  2020      1    1         2   
2 2020-01-02        0.0         0.0        NaN  2020      1    2         3   
3 2020-01-03        0.0         0.0        NaN  2020      1    3         4   
4 2020-01-04        0.0         0.0        NaN  2020      1    4         5   

   total_cases  total_deaths  total_tests location continent  population  \
0          0.0           0.0          0.0    Italy    Europe  60461828.0   
1          0.0           0.0          0.0    Italy    Europe  60461828.0   
2          0.0           0.0          0.0    Italy    Europe  60461828.0   
3          0.0           0.0          0.0    Italy    Europe  60461828.0   
4          0.0           0.0          0.0    Italy    Europe  60461828.0   

   life_expectancy  hospital_beds_per_thousand  gdp_per_capita  
0            83.51                        3.18       35220.084  
1            83.51                        3.18       35220.084  
2            83.51                        3.18       35220.084  
3            83.51                        3.18       35220.084  
4            83.51                        3.18       35220.084

The location data for Italy is appended to each row within `covid_df`. If the `covid_df` data frame contained data for multiple locations, then the respective country's location data would be appended for each row.

We can now calculate metrics like cases per million, deaths per million, and tests per million.

In [102]:
merge_df['cases_per_million'] = merge_df.total_cases * 1e6 / merge_df.population

In [103]:
merge_df['deaths_per_million'] = merge_df.total_deaths * 1e6 / merge_df.population

In [104]:
merge_df['tests_per_million'] = merge_df.total_tests * 1e6 / merge_df.population

In [106]:
merge_df.tail()

date  new_cases  new_deaths  new_tests  year  month  day  weekdays  \
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30         6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31         0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1         1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2         2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3         3   

     total_cases  total_deaths  total_tests location continent  population  \
243     267298.5       35473.0      35473.0    Italy    Europe  60461828.0   
244     268663.5       35477.0      35477.0    Italy    Europe  60461828.0   
245     269659.5       35483.0      35483.0    Italy    Europe  60461828.0   
246     270634.5       35491.0      35491.0    Italy    Europe  60461828.0   
247     271960.5       35497.0      35497.0    Italy    Europe  60461828.0   

     life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
243            83.51                        3.18       35220.084   
244            83.51                        3.18       35220.084   
245            83.51                        3.18       35220.084   
246            83.51                        3.18       35220.084   
247            83.51                        3.18       35220.084   

     cases_per_million  deaths_per_million  tests_per_million  
243        4420.946386          586.700753         586.700753  
244        4443.522614          586.766910         586.766910  
245        4459.995818          586.866146         586.866146  
246        4476.121695          586.998461         586.998461  
247        4498.052887          587.097697         587.097697

## Writing data back to files

After completing your analysis and adding new columns, you should write the results back to a file. Otherwise, the data will be lost when the Jupyter notebook shuts down. Before writing to file, let us first create a data frame containing just the columns we wish to record.

In [107]:
result_df = merge_df[['date',
                       'new_cases', 
                       'total_cases', 
                       'new_deaths', 
                       'total_deaths', 
                       'new_tests', 
                       'total_tests', 
                       'cases_per_million', 
                       'deaths_per_million', 
                       'tests_per_million']]

In [108]:
result_df.tail()

date  new_cases  total_cases  new_deaths  total_deaths  new_tests  \
243 2020-08-30     1444.0     267298.5         1.0       35473.0    53541.0   
244 2020-08-31     1365.0     268663.5         4.0       35477.0    42583.0   
245 2020-09-01      996.0     269659.5         6.0       35483.0    54395.0   
246 2020-09-02      975.0     270634.5         8.0       35491.0        NaN   
247 2020-09-03     1326.0     271960.5         6.0       35497.0        NaN   

     total_tests  cases_per_million  deaths_per_million  tests_per_million  
243      35473.0        4420.946386          586.700753         586.700753  
244      35477.0        4443.522614          586.766910         586.766910  
245      35483.0        4459.995818          586.866146         586.866146  
246      35491.0        4476.121695          586.998461         586.998461  
247      35497.0        4498.052887          587.097697         587.097697

To write the data from the data frame into a file, we can use the `to_csv` function. 

In [109]:
result_df.to_csv('result.csv', index = None)